torch/csrc/onnx/export.cpp

```diff
@@ -256,18 +256,46 @@ static void addAttribute(onnx::NodeProto * n_p, jit::Node * n, jit::Symbol name)
       break;
     case AttributeKind::g: {
       auto g = attr->mutable_g();
-      encodeGraph(g, n->g(name), {});
+      encodeGraph(g, n->g(name), {}, {}, {});
     } break;
     case AttributeKind::gs:
       for(auto & v : n->gs(name)) {
         auto g = attr->add_graphs();
-        encodeGraph(g, v, {});
+        encodeGraph(g, v, {}, {}, {});
       }
       break;
   }
 }
 
-static void encodeGraph(onnx::GraphProto * p_g, std::shared_ptr<Graph> & g, const std::vector<at::Tensor> & initializers) {
+
+static void encodeGraph(onnx::GraphProto * p_g, std::shared_ptr<Graph> & g,
+  const std::vector<std::string> & input_names,
+  const std::vector<std::string> & output_names,
+  const std::vector<at::Tensor> & initializers) {
+
+  std::unordered_map<Node *, std::string> node_names;
+  auto initNames = [&](const node_list & nodes, const std::vector<std::string> & names) {
+    JIT_ASSERT(names.size() <= nodes.size());
+    size_t idx = nodes.size() - names.size();
+    for (auto & name : names) {
+      auto node = nodes.at(idx++);
+      auto it = node_names.find(node);
+      // rarely, output nodes may be repeated
+      // if this happens, we just take the first name we find for it as its name
+      if(it == node_names.end()) {
+        node_names[node] = name;
+      }
+    }
+  };
+  initNames(g->inputs(), input_names);
+  initNames(g->outputs(), output_names);
+  auto node_name = [&](Node * n) {
+    auto it = node_names.find(n);
+    if(it != node_names.end())
+      return it->second;
+    return "$" + n->uniqueName();
```

Comment:
Can you suggest an alternative?

Reply:
Well, there are a few possibilities.

1. Reserve identifiers which start with `[0-9]`. Now you don't need to rename the numeric identifiers with a leading dollar. This is what LLVM does, btw: http://llvm.org/docs/LangRef.html#identifiers
2. Fix up the ONNX printer to not prepend the output with `%`. Perhaps if we want named identifiers to still display with a prefix, we can test if the first character is a `$` or not, and prepend `%` if it is.

Reply:
I think I will just enforce that input_names/output_names must be a valid C identifier. That makes sure it is not equal to a number-based thing, and we can make our uniqueName insert a non-identifier character to prevent overlap with debugNames as well.